pkg/store/multitsdb.go

```diff
@@ -0,0 +1,284 @@
+// exceed the requested time bounds.
+func (s *MultiTSDBStore) Series(r *storepb.SeriesRequest, srv storepb.Store_SeriesServer) error {
+	stores := s.tsdbStores()
+	if len(stores) == 0 {
+		return nil
+	}
+
+	var (
+		g, gctx = errgroup.WithContext(srv.Context())
+
+		// Allow to buffer max 10 series response.
+		// Each might be quite large (multi chunk long series given by sidecar).
+		respSender, respRecv, closeFn = newRespCh(gctx, 10)
+	)
+
+	g.Go(func() error {
+		var (
+			seriesSet []storepb.SeriesSet
+			wg        = &sync.WaitGroup{}
+		)
+
+		defer func() {
+			wg.Wait()
+			closeFn()
+		}()
+
+		for tenant, store := range stores {
+			store := store
+			seriesCtx, closeSeries := context.WithCancel(gctx)
+			seriesCtx = grpc_opentracing.ClientAddContextTags(seriesCtx, opentracing.Tags{
+				"tenant": tenant,
+			})
+			defer closeSeries()
+			ss := newSeriesSetServer(seriesCtx, respSender)
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				ss.Run(store, r)
+			}()
+
+			seriesSet = append(seriesSet, ss)
+		}
+
+		mergedSet := storepb.MergeSeriesSets(seriesSet...)
+		for mergedSet.Next() {
+			var series storepb.Series
+			series.Labels, series.Chunks = mergedSet.At()
+			respSender.send(storepb.NewSeriesResponse(&series))
+		}
+		return mergedSet.Err()
+	})
+
+	for resp := range respRecv {
+		if err := srv.Send(resp); err != nil {
+			return status.Error(codes.Unknown, errors.Wrap(err, "send series response").Error())
+		}
+	}
+
+	if err := g.Wait(); err != nil {
+		level.Error(s.logger).Log("err", err)
+		return err
+	}
+	return nil
+}
+
+// LabelNames returns all known label names.
+func (s *MultiTSDBStore) LabelNames(ctx context.Context, req *storepb.LabelNamesRequest) (*storepb.LabelNamesResponse, error) {
+	names := map[string]struct{}{}
+	warnings := map[string]struct{}{}
+
+	stores := s.tsdbStores()
+	for _, store := range stores {
+		r, err := store.LabelNames(ctx, req)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, l := range r.Names {
+			names[l] = struct{}{}
+		}
+
+		for _, l := range r.Warnings {
+			warnings[l] = struct{}{}
+		}
+	}
+
+	return &storepb.LabelNamesResponse{
+		Names:    keys(names),
+		Warnings: keys(warnings),
+	}, nil
+}
+
+func keys(m map[string]struct{}) []string {
+	res := make([]string, 0, len(m))
+	for k := range m {
+		res = append(res, k)
+	}
+
+	return res
+}
+
+// LabelValues returns all known label values for a given label name.
+func (s *MultiTSDBStore) LabelValues(ctx context.Context, req *storepb.LabelValuesRequest) (*storepb.LabelValuesResponse, error) {
+	values := map[string]struct{}{}
+	warnings := map[string]struct{}{}
+
+	stores := s.tsdbStores()
+	for _, store := range stores {
+		r, err := store.LabelValues(ctx, req)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, l := range r.Values {
+			values[l] = struct{}{}
+		}
+
+		for _, l := range r.Warnings {
+			warnings[l] = struct{}{}
```

Comment:
do we want really deduplicate warnings :thinking:  same for names

Reply:
Duplicate labelnames in the response would be more than confusing. Same with warnings, it doesn't seem helpful to the user if the same warning shows up multiple times.

Reply:
I disagree, they relate to different DBs, no?